src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts

```diff
@@ -361,35 +363,82 @@ export class CommandDetectionCapability extends Disposable implements ICommandDe
 		this._logService.debug('CommandDetectionCapability#handleCommandStart', this._currentCommand.commandStartX, this._currentCommand.commandStartMarker?.line);
 	}
 
-	private _handleCommandStartWindows(): void {
+	private async _handleCommandStartWindows(): Promise<void> {
+		if (this._windowsPromptPollingInProcess) {
+			this._windowsPromptPollingInProcess = false;
+		}
+		this._commandStartedWindowsBarrier = new Barrier();
 		this._currentCommand.commandStartX = this._terminal.buffer.active.cursorX;
 
 		// On Windows track all cursor movements after the command start sequence
 		this._commandMarkers.length = 0;
-		// HACK: Fire command started on the following frame on Windows to allow the cursor
-		// position to update as conpty often prints the sequence on a different line to the
-		// actual line the command started on.
-		timeout(0).then(() => {
-			if (!this._currentCommand.commandExecutedMarker) {
-				this._onCursorMoveListener = this._terminal.onCursorMove(() => {
-					if (this._commandMarkers.length === 0 || this._commandMarkers[this._commandMarkers.length - 1].line !== this._terminal.buffer.active.cursorY) {
-						const marker = this._terminal.registerMarker(0);
-						if (marker) {
-							this._commandMarkers.push(marker);
-						}
+
+		// Conpty could have the wrong cursor position at this point.
+		if (!this._cursorOnNextLine() || !this._cursorLineLooksLikeWindowsPrompt()) {
+			this._windowsPromptPollingInProcess = true;
+			// Poll for 200ms until the cursor position is correct.
+			let i = 0;
+			for (; i < 20; i++) {
+				await timeout(10);
```

Comment:
Does this need to handle cases where the `this` object is disposed of while the timeouts are still firing?